Security.md

```diff
@@ -0,0 +1,122 @@
+  with limited exceptions.
+
+Additionally, each module is subject to the security policies of its embedding.
+Within a [web browser](Web.md), this includes restrictions on information flow
+through [same-origin policy][]. On a [non-web](NonWeb.md) platform, this could
+include the POSIX security model.
+
+## Developers
+
+The design of WebAssembly promotes safe programs by eliminating dangerous
+features from its execution semantics, while maintaining compatibility with
+programs written for [C/C++](CandC++.md).
+
+Modules must declare all accessible functions and their associated types
+at load time, even when [dynamic linking](DynamicLinking.md) is used. This
+allows implicit enforcement of [control-flow integrity][] (CFI) through
+structured control-flow. Since compiled code is immutable and not observable at
+runtime, WebAssembly programs are protected from control flow hijacking attacks.
+  * [Function calls](AstSemantics.md#calls) must specify the index of a target
+  that corresponds to a valid entry in the
+  [function index space](Modules.md#function-index-space) or
+  [table index space](Modules.md#table-index-space)
+  * [Indirect function calls](Rationale.md#indirect-calls) are subject to a type
+  signature check at runtime; the type signature of the selected indirect
+  function must match the type signature specified at the call site.
+  * A shadow stack is used to maintain a trusted call stack that is invulnerable
+  to buffer overflows and immutable outside the runtime, ensuring safe function
+  returns.
+  * [Branches](AstSemantics.md#branches-and-nesting) must point to valid
+  destinations within the enclosing function.
+
+Variables in C/C++ can be lowered to two different primitives in WebAssembly,
+depending on their scope. [Local variables](AstSemantics.md#local-variables)
+with fixed scope and [global variables](AstSemantics.md#global-variables) are
+represented as fixed-type values stored by index. The former are initialized
+to zero by default and are stored in the protected shadow stack, whereas
+the latter are located in the [global index space](Modules.md#global-index-space)
+and can be imported from external modules. Local variables with
+[unclear static scope](Rationale.md#locals) (e.g. are used by the address-of
+operator, or are of type `struct` and returned by value) are stored in a separate
+user-addressable stack in [linear memory](AstSemantics.md#linear-memory) at
+compile time. This is an isolated memory region with fixed maximum size that is
+zero initialized by default. References to this memory are computed with
+infinite precision to avoid wrapping and simplify bounds checking. In the future,
+support for [multiple linear memory sections](Modules.md#linear-memory-section) and
+[finer-grained memory operations](FutureFeatures.md#finer-grained-control-over-memory)
+(e.g. shared memory, page protection, large pages, etc.) will be implemented.
```

Comment:
We don't really need to detail how every future feature could be secured at this point in time. This is the first PR in a series, I'm fine with leaving some work for later.